PRCS/prc-4.md

```diff
@@ -0,0 +1,138 @@
+    /// If msg.value is more than the sum of orders' prices, it SHOULD refund the difference back to msg.sender.
+    function fillSellOrders(uint256[] memory orderIds) external payable;
+
+    /// @notice Cancels the sell order specified by `orderId`, making it unfillable.
+    /// @dev Reverts if the msg.sender is not the order's seller.
+    /// MUST change the `active` parameter for the specified order to `false`.
+    /// MUST emit `OrderCancelled` event.
+    function cancelSellOrder(uint256 orderId) external;
+}
+```
+
+## Rationale
+
+It's not an AMM. In a typical AMM dex, there are liquidity pools and you trade against these pools. This is unsuitable for this use-case because a contract on base chain has no means of knowing the state of the game chain so the contract cannot automate market making. Instead, buyers have to be sending tokens directly to the sellers.
+
+### The idea:
+
+1. Smart contract on base chain is created to facilitate trading game asset. This contract allows the following:
+    * Function to create a sell order. A sell order persists and has the following properties: 
+      * `assetAmount` - amount of the game asset the seller is selling,
+      * `price` - the price in native gas tokens of the base chain the seller is requesting,
+      * `seller` - the address of the seller,
+      * `active` - signals whether or not the sell order is active (meaning it has been cancelled or filled)
+    * Function to cancel a sell order.
+    * Function to fill a sell order (in other words - buy). There is no "buy order" that persists, and rather the buy function directly transfers the value specified by price defined in the orders to the sellers. Buying marks all the sell orders it fulfills as inactive by changing the value of `active` flag to `false`.
+2. The game chain monitors this contract using Paima Primitives and exposes an API to query its state.
+
+The contract on the base chain has no way of knowing if somebody who makes a sell order really has that amount of game assets in their account. Rather,
+
+1. When somebody creates a sell order: The game chain sees it and checks if the user has enough of the asset in their account.
+    * If yes, the amount is locked so they cannot spend it in-game (unless they cancel the order).
+    * If not, the order is marked as invalid (to differentiate it from the case where it doesn't know the order exists yet).
+2. When somebody wants to buy, they specify the orders they wish to purchase by ID. 
+    * The responsibility of querying the API of game chain to check the validity of the sell order is left up to the front-end providers, as well as the presentation of the possible sell orders.
+3. The user then makes a smart contract call that fulfills all the orders and sends the right amount to the corresponding accounts in a single transaction (atomic).
+4. The game chain monitors the successfully fulfilled orders on the base chain and transfers the assets between accounts.
+
+**Note:**
```

Comment:
With partial fills, why would we have to discard the old order and create a new order? I understand that you're thinking in terms of UTxO model, but is this really necessary? We can just have the order track the amount and have it change with partial fills.
Example: There is sell order for 1000 TGOLD. I submit a buy for 800 TGOLD but somebody else submits a buy of 600 TGOLD with higher gas fee, so by the time my transaction is being processed, 600 was already bought and so the sell order would only have 400 TGOLD left. I would much rather have my buy order fill for the 400 TGOLD (because it was a good price per token) than it to revert and me not getting anything and losing money on failed transaction gas fee.

Reply:
> With partial fills, why would we have to discard the old order and create a new order

Yes, as I mentioned in my post above we can just do the option `orderID is constant`. I initially thought we would have to create new orders every time for the game to react based off of it, but I realized that wasn't necessary since the RPC endpoint I mentioned is enough for the game endpoint to return the right state without having to see the tx on Arbitrum